src/esp/sim/Simulator.cpp

```diff
@@ -589,6 +600,156 @@ void Simulator::reconfigureReplayManager(bool enableGfxReplaySave) {
+        "Simulator::computeShadowMaps: To compute the shadow map, the light"
+            << iLight << "should be in `global` mode.", );
+    CORRADE_ASSERT(lightingSetup[iLight].vector.w() == 1,
+                   "Simulator::computeShadowMaps: Only point light shadow is "
+                   "supported. However, the light"
+                       << iLight << "is a directional light.", );
+    Mn::Vector3 lightPos = Mn::Vector3{lightingSetup[iLight].vector.xyz()};
+
+    Mn::ResourceKey key(Corrade::Utility::formatString(
+        assets::ResourceManager::SHADOW_MAP_KEY_TEMPLATE, activeSceneID_,
+        iLight));
+
+    // insert the key to the data base
+    keys.push_back(key);
+
+    // create the resource if there is not one
+    Mn::Resource<gfx::CubeMap> pointShadowMap =
+        shadowManager.get<gfx::CubeMap>(key);
+    if (!pointShadowMap) {
+      shadowManager.set<gfx::CubeMap>(
+          pointShadowMap.key(),
+          new gfx::CubeMap{
+              shadowMapSize,
+              {gfx::CubeMap::Flag::VarianceShadowMapTexture |
+               // gfx::CubeMap::Flag::ColorTexture | // for future visualization
+               gfx::CubeMap::Flag::AutoBuildMipmap}},
+          Mn::ResourceDataState::Final, Mn::ResourcePolicy::Resident);
+
+      CORRADE_INTERNAL_ASSERT(pointShadowMap && pointShadowMap.key() == key);
+    }
+    if (pointShadowMap->getCubeMapSize() != shadowMapSize) {
+      pointShadowMap->reset(shadowMapSize);
+    }
+
+    // setup a CubeMapCamera in the root of scene graph, and set its position to
+    // light global position
+    lightNode.setTranslation(lightPos);
+
+    camera.setProjectionMatrix(shadowMapSize,   // width of the square
+                               lightNearPlane,  // near plane
+                               lightFarPlane);  // far plane
+    pointShadowMap->renderToTexture(camera, sg, shadowMapDrawableGroupName,
+                                    {gfx::RenderCamera::Flag::FrustumCulling |
+                                     gfx::RenderCamera::Flag::ClearDepth});
+
+    Mn::ResourceKey helperKey("helper-shadow-cubemap");
+    Mn::Resource<gfx::CubeMap> helperShadowMap =
+        shadowManager.get<gfx::CubeMap>(helperKey);
+    if (!helperShadowMap) {
+      shadowManager.set<gfx::CubeMap>(
+          helperShadowMap.key(),
+          new gfx::CubeMap{shadowMapSize,
+                           {gfx::CubeMap::Flag::VarianceShadowMapTexture |
+                            gfx::CubeMap::Flag::AutoBuildMipmap}},
+          Mn::ResourceDataState::Final, Mn::ResourcePolicy::ReferenceCounted);
+
+      CORRADE_INTERNAL_ASSERT(helperShadowMap &&
+                              helperShadowMap.key() == helperKey);
+    }
+
+    // for VSM only !!!
+    renderer_->applyGaussianFiltering(
+        *pointShadowMap, *helperShadowMap,
+        gfx::CubeMap::TextureType::VarianceShadowMap);
+  }
+}
+
+void Simulator::setShadowMapsToDrawables() {
```

Comment:
Would it be feasible and possibly appropriate to put some or all of these 3 functions that directly access the ShadowMap constructs (updateShadowMapDrawableGroup, computeShadowMaps, setShadowMapsToDrawables) in ResourceManager, so we do not have to provide public access to these constructs from outside RM?  We could pass in refs to current scenegraph and renderer (where appropriate). 

Just a thought :)

Reply:
Yeah, I see your point.
My opinion: `Simulator` class is like a driver to glue the other library modules.`Shadows` are in the "resource manager". But "Drawables" belong to the `SceneGraph` which is in another module. `setShadowMapsToDrawables`, as well as the other 2 functions, will work on these 2 different modules. So I put them in the `Simulator` to avoid complicated parameter passing.

Besides `Simulator` will have to provide such APIs for the external users (e.g., `viewer`) to construct shadow maps.

Reply:
Fair enough.Thanks for responding :)